chia/wallet/nft_wallet/nft_off_chain.py

```diff
@@ -0,0 +1,108 @@
+            async with session.get(str(uri, "utf-8")) as response:
+                if response.status == 200:
+                    text = await response.text()
+                    if ignore_size_limit or sys.getsizeof(text) / 1024 / 1024 < 1:
+                        return text
+    return None
+
+
+def read_off_chain_metadata(nft_coin_info: NFTCoinInfo, cache_path: Optional[str] = None) -> Optional[str]:
+    # Read metadata from disk cache
+    file_name = get_cached_filename(nft_coin_info.nft_id, cache_path)
+    if not file_name.exists():
+        return None
+    try:
+        text = file_name.read_text()
+        if verify_metadata(text, nft_coin_info.full_puzzle):
+            return text
+        else:
+            return None
+    except Exception:
+        log.exception(f"Cannot load cached metadata of {nft_coin_info.nft_id.hex()}.")
+        return None
+
+
+def delete_off_chain_metadata(nft_id: bytes32, cache_path: Optional[str] = None) -> None:
+    file_name = get_cached_filename(nft_id, cache_path)
+    if file_name.exists():
+        file_name.unlink()
+        log.info(f"Deleted off-chain metadata of {nft_id.hex()}")
+
+
+def write_off_chain_metadata(cache_path: Optional[str], nft_id: bytes32, metadata: str) -> None:
+    file_name = get_cached_filename(nft_id, cache_path)
+    file_name.write_text(metadata)
+
+
+async def get_off_chain_metadata(
+    nft_coin_info: NFTCoinInfo, cache_path: Optional[str] = None, ignore_size_limit: bool = False
+) -> Optional[str]:
+    try:
+        # Check if the metadata is in disk cache
+        metadata = read_off_chain_metadata(nft_coin_info, cache_path)
+        if metadata is not None:
+            return metadata
+        log.debug(f"{nft_coin_info.nft_id.hex()} is not in cache, downloading now ...")
+        metadata = await fetch_off_chain_metadata(nft_coin_info, ignore_size_limit)
+        if metadata is None:
+            log.error(f"Cannot find off-chain metadata of {nft_coin_info.nft_id.hex()}.")
+            return None
+        write_off_chain_metadata(cache_path, nft_coin_info.nft_id, metadata)
+        log.info(f"Loaded off-chain metadata of {nft_coin_info.nft_id.hex()}")
+        return metadata
+    except Exception:
+        log.exception(f"Cannot get off-chain metadata of {nft_coin_info.nft_id.hex()}.")
+        return None
+
+
+def verify_metadata(metadata: str, full_puzzle: Program) -> bool:
+    uncurried_nft: Optional[UncurriedNFT] = UncurriedNFT.uncurry(*full_puzzle.uncurry())
+    if uncurried_nft is None:
+        return False
+    if uncurried_nft.meta_hash.as_python().hex() == std_hash(str.encode(metadata)).hex():
+        return True
+    else:
+        log.error(
+            f"Expect metadata hash: {uncurried_nft.meta_hash.as_python().hex()},"
+            f" actual metadata hash {std_hash(str.encode(metadata)).hex()}"
+        )
+        return False
+
+
+def get_cached_filename(nft_id: bytes32, cache_path: Optional[str] = None) -> Path:
```

Comment:
I see that `cache_path` being a string is propagated from existing code so I won't request a change on it here.  But, be aware that such deserialization should be done at the boundaries rather than pushing the serialized form all throughout the code.